pkg/server/commands/reverseexpand/reverse_expand.go

```diff
@@ -267,6 +267,16 @@ func (c *ReverseExpandQuery) execute(
 		sourceUserObj = val.ObjectRelation.GetObject()
 		sourceUserRef = typesystem.DirectRelationReference(sourceUserType, val.ObjectRelation.GetRelation())
 
+		sourceUserRel := val.ObjectRelation.GetRelation()
+
+		if req.edge == nil {
```

Comment:
**question**. With this line we are only consider this "fact" as true on the first (base) call of the recursion. Whereas in Check, you are treating it as true on the base case and every subsequent recursion call. Is this intentional?

Reply:
**Another question:** Is this check required? Couldn't you have a self-referencing tuple as a result of another expansion or itself expand further?

Ex:

```
model
  schema 1.1
type user
  relations
    define editor: [user]
    define viewer: [user] or editor
```
```
Check(user:1#editor@user:1#editor) // True
Check(user:1#editor@user:1#viewer) // ??? 
```

Another case to ponder:
```
model
  schema 1.1
type user
  relations
    define blocked: [user]
    define viewer: [user] but not blocked
```
```
user:1#blocked@user:1
```
```
Check(user:1#viewer@user:1#viewer) // ??
```

Reply:
But we don't have a top-level ResolveCheck. ResolveCheck is recursive

Reply:
@miparnisari could you be more specific with what you are wanting to see? Are you suggesting that we move this up to the `Execute` function instead of inside of the `execute` function? If so, I can do that, but it's functionally the same. `execute` where `req.edge==nil` only happens on the first invocation, because any subsequent recursive call will have a non-nil edge.

Reply:
I'm not sure what is the correct approach here, whether it's okay to do the check on the top-level call only, or we should also do it in each sub-problem. I haven't thought about the implications. What i'm pointing out is that in your PR this isn't consistent. 

Since ResolveCheck is for both the top-level call and the sub-problems, until we figure out the implications, IReverseExpand should behave the same and we should get rid of `if req.edge == nil` check

Reply:
> whether it's okay to do the check on the top-level call only, or we should also do it in each sub-problem. 

That's what I was trying to get at above. It would be odd if a reflexive userset only returned itself at the top-level call but not for all other subproblems. @jon-whit is that even possible?

Reply:
Discussed on Zoom. Let's remove `if req.edge == nil` :)

Reply:
We talked about this offline, but just reposting for broader visibility and historical context.

I've added some tests that exhibit some of the behaviors around the questions asked in this thread. The `tl;dr` is all based around set theory. If every subject of some subject set would also be guaranteed to be a subset of the other, then we should be considering the subgraphs as equivalent subgraphs. As just one simple example,

```
type user

type document
  relations
    define writer: [user]
    define editor: [user]
    define viewer: writer or editor
```

`Check(document:1#viewer@document:1#writer)` --> `{allowed: true}`, because every subject who is a subject of the set `document:1#writer` is also a subject of the set `document:1#viewer`. Similarly `Check(document:1#viewer@document:1#writer)` --> `{allowed: true}`.

However, this isn't necessarily true in all cases, for example,

```
type user

type document
  relations
    define allowed: [user]
    define viewer: [user] and allowed
```

`Check(document:1#viewer@document:1#allowed)` --> {allowed: false} because not all subjects of `document:1#allowed` are also part of the set `document:1#viewer`.